serde-generate/src/typescript.rs

```diff
@@ -0,0 +1,857 @@
+        let mut emitter = TypeScriptEmitter {
+            out: IndentedWriter::new(out, IndentConfig::Space(2)),
+            generator: self,
+        };
+
+        emitter.output_preamble()?;
+
+        for (name, format) in registry {
+            emitter.output_container(name, format)?;
+        }
+
+        if self.config.serialization {
+            emitter.output_helpers(registry)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<'a, T> TypeScriptEmitter<'a, T>
+where
+    T: Write,
+{
+    fn output_preamble(&mut self) -> Result<()> {
+        writeln!(
+            self.out,
+            "import {{BigNumber}} from '@ethersproject/bignumber';"
+        )?;
+        writeln!(
+            self.out,
+            "import {{Int64LE, Uint64LE}} from 'int64-buffer';\n"
+        )?;
+        writeln!(self.out, "import bytes from '@ethersproject/bytes';\n")?;
+        writeln!(
+            self.out,
+            "import {{ Serializer }} from '../serde/serializer';\n"
+        )?;
+        writeln!(
+            self.out,
+            "import {{ Deserializer }} from '../serde/deserializer';\n"
+        )?;
+        for namespace in self.generator.namespaces_to_import.iter() {
+            writeln!(
+                self.out,
+                "import * as {} from '../{}';\n",
+                namespace.to_camel_case(),
+                namespace
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn quote_qualified_name(&self, name: &str) -> String {
+        self.generator
+            .external_qualified_names
+            .get(name)
+            .cloned()
+            .unwrap_or_else(|| name.to_string())
+    }
+
+    fn output_comment(&mut self, name: &str) -> std::io::Result<()> {
+        let mut path = Vec::new();
+        path.push(name.to_string());
+        if let Some(doc) = self.generator.config.comments.get(&path) {
+            let text = textwrap::indent(doc, " * ").replace("\n\n", "\n *\n");
+            writeln!(self.out, "/**\n{} */", text)?;
+        }
+        Ok(())
+    }
+
+    fn quote_type(&self, format: &Format) -> String {
+        use Format::*;
+        match format {
+            TypeName(x) => self.quote_qualified_name(x),
+            Unit => "void".into(),
+            Bool => "boolean".into(),
+            I8 => "number".into(),
+            I16 => "number".into(),
+            I32 => "number".into(),
+            I64 => "Int64LE".into(),
+            I128 => "BigNumber".into(),
+            U8 => "number".into(),
+            U16 => "number".into(),
+            U32 => "number".into(),
+            U64 => "Uint64LE".into(),
+            U128 => "BigNumber".into(),
+            F32 => "number".into(),
+            F64 => "number".into(),
+            Char => "string".into(),
+            Str => "string".into(),
+            Bytes => "Uint8Array".into(),
+
+            Option(format) => format!("Optional{}", self.quote_type(format)),
```

Comment:
Using a generic type alias would be better IMHO. See other comment.

(Incidentally for what you intended to do, it should be `common::mangle_type` here, not `quote_type`.)

Reply:
will be changed.